add ignored_states to FSDP/fully_shard

zhaojuanmao opened the pull request:
Add 'ignored_states' that accepts either a list of ignored_parameters or a list of nn modules for FSDP model wrapper and fully_shard composable APIs, it is recommended to use 'ignored_states' over 'ignored_modules' moving forward

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102056

zhaojuanmao committed:
add ignored_states to FSDP/fully_shard

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -96,7 +99,7 @@ def fully_shard(
     _register_root_pre_forward_hook(state, module)  # prepend last
     for submodule in module.modules():
         if (
-            submodule not in state._ignored_modules
```
Users may pass 'ignored_parameters' instead of 'ignored_modules', in this case, it is not proper to use 'state._ignored_modules' here.

awgu replied:
If we ignore _some_ parameters from a module, then what should the semantics be? Should we still annotate it with `fully_shard`? Should we only annotate it with `fully_shard` if we ignored _all_ of the module's parameters (`recurse=False`)?

zhaojuanmao replied:
good catch! I think we should insert the state for the sharded module even it has partial managed parameter and partial ignored parameters. 

As commented below, probably we will just use '_fully_sharded_module_to_handles' to decide whether we should insert states for sharded modules or not

awgu commented on torch/distributed/_composable/fully_shard.py:
```diff
@@ -96,7 +99,7 @@ def fully_shard(
     for submodule in module.modules():
         if (
-            submodule not in state._ignored_modules
+            submodule.__class__.__name__ in state._fully_sharded_module_names
```
My main concern is whether using `.__class__.name__` is too brittle. 

In my understanding, your PR uses `_fully_sharded_module_names` only for the `fully_shard` path to determine for which submodules you should call `_insert_module_state(submodule, state)`.

1. If we ignore parameters from one `nn.Linear`, then will we skip calling `_insert_module_state()` for _all_ `nn.Linear`? (because `"Linear"` will be in `_fully_sharded_module_names`)
2. If we ignore parameters from one `nn.Linear`, then will we skip calling `_insert_module_state()` for _all_ module classes called `Linear` (even if not from `nn`)? (because they all have `.__class__.__name__ == "Linear"`)

```
>>> lin1 = nn.Linear(3, 3)  # use built-in from nn
>>> class Linear(nn.Module):  # define own
>>>     def __init__(self):
>>>         super().__init__()
>>> lin2 = Linear()
>>> lin1.__class__.__name__
'Linear'
>>> lin2.__class__.__name__
'Linear'
```

The first question is a more important concern since it seems to affect correctness directly.

zhaojuanmao replied:
That is a good point! I think the proper way is to use '_fully_sharded_module_to_handles' instead of '_fully_sharded_module_names' that is originally written for debugging (even though it is not ideal for debugging as well). 

One minor issue for using '_fully_sharded_module_to_handles' is that: the fully_sharded_module without parameters/handles will not have a fsdp state, it should be fine... as it does not have parameters to be managed by fsdp any way? I'm just worry about whether some edge cases I may miss?

zhaojuanmao committed:
Update on "add ignored_states to FSDP/fully_shard"

Add 'ignored_states' that accepts either a list of ignored_parameters or a list of nn modules for FSDP model wrapper and fully_shard composable APIs, it is recommended to use 'ignored_states' over 'ignored_modules' moving forward




[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -166,25 +192,48 @@ def test_ignored_modules_nested(self):
         error and still ignores non-FSDP modules' parameters."""
         self.run_subtests(
-            {"use_orig_params": [False, True], "ignore_modules": [True, False]},
+            {
```
Are we running `composable=True` twice because it ignores `use_orig_params=True` vs. `False`?

awgu commented on test/distributed/fsdp/test_fsdp_ignored_modules.py:
```diff
@@ -212,24 +265,31 @@ def _test_ignored_modules_nested(self, use_orig_params: bool, ignore_modules: bo
         top-level module itself errors."""
         model = Model().cuda()
-        model.layer1 = FSDP(model.layer1)
+        wrap_cls = FSDP
+        if composable:
+            wrap_cls = fully_shard
```
nit: haha 
```suggestion
        wrap_cls = fully_shard if composable else FSDP
```

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -659,7 +668,9 @@ def _get_ignored_modules(
         if not isinstance(module, torch.nn.Module):
             raise TypeError(msg_prefix + f"but got an iterable with {type(module)}")
-        if isinstance(module, fsdp_file.FullyShardedDataParallel):
+        if isinstance(
+            module, fsdp_file.FullyShardedDataParallel
+        ) or _get_module_fsdp_state(module):
```
Since this is only called once during initialization, I think we can simplify this to just:
```
if _get_module_fsdp_state(module):
```
?

I am thinking the reason is that for wrapper path, we still have `_get_module_fsdp_state(module) is not None`.

awgu reviewed (approved):
I think this change makes sense!

zhaojuanmao committed:
Update on "add ignored_states to FSDP/fully_shard"

Add 'ignored_states' that accepts either a list of ignored_parameters or a list of nn modules for FSDP model wrapper and fully_shard composable APIs, it is recommended to use 'ignored_states' over 'ignored_modules' moving forward




[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot merge
